src/checkers/inference/solver/backend/z3backend/Z3BitVectorSerializer.java

```diff
@@ -0,0 +1,312 @@
+        return equalityVariableCombos.accept(constraint.getFirst(), constraint.getSecond(), constraint);
+    }
+
+    @Override
+    public BoolExpr serialize(InequalityConstraint constraint) {
+        // TODO Can be supported.
+        return EMPTY_VALUE;
+    }
+
+    public BitVecExpr serializeVarSlot(VariableSlot slot) {
+        int slotId = slot.getId();
+
+        if (serializedSlots.containsKey(slotId)) {
+            return serializedSlots.get(slotId);
+        }
+
+        BitVecExpr bitVector = context.mkBVConst(String.valueOf(slot.getId()),
+                z3BitVectorCodec.getFixedBitVectorSize());
+        serializedSlots.put(slotId, bitVector);
+
+        return bitVector;
+    }
+
+    public BitVecExpr serializeConstantSlot(ConstantSlot slot) {
+        int slotId = slot.getId();
+
+        if (serializedSlots.containsKey(slotId)) {
+            return serializedSlots.get(slotId);
+        }
+
+        BigInteger numeralValue = z3BitVectorCodec.encodeConstantAM(slot.getValue());
+
+        BitVecNum bitVecNum = context.mkBV(numeralValue.toString(), z3BitVectorCodec.getFixedBitVectorSize());
+        serializedSlots.put(slotId, bitVecNum);
+
+        return bitVecNum;
+    }
+
+    @Override
+    public BitVecExpr serialize(VariableSlot slot) {
+        return serializeVarSlot(slot);
+    }
+
+    @Override
+    public BitVecExpr serialize(ConstantSlot slot) {
+        return serializeConstantSlot(slot);
+    }
+
+    @Override
+    public BitVecExpr serialize(ExistentialVariableSlot slot) {
+        return serializeVarSlot(slot);
+    }
+
+    @Override
+    public BitVecExpr serialize(RefinementVariableSlot slot) {
+        return serializeVarSlot(slot);
+    }
+
+    @Override
+    public BitVecExpr serialize(CombVariableSlot slot) {
+        return serializeVarSlot(slot);
+    }
+
+    @Override
+    public BoolExpr serialize(ExistentialConstraint constraint) {
+        // Not supported.
+        return EMPTY_VALUE;
+    }
+
+    @Override
+    public BoolExpr serialize(ComparableConstraint comparableConstraint) {
+        // Not supported.
+        return EMPTY_VALUE;
+    }
+
+    @Override
+    public BoolExpr serialize(CombineConstraint combineConstraint) {
+        // Not supported.
+        return EMPTY_VALUE;
+    }
+
+    @Override
+    /**
+     * Return an equality constriant between variable and constant goal.
+     * The caller should add the serialized constraint with soft option.
+     */
+    public BoolExpr serialize(PreferenceConstraint preferenceConstraint) {
+        VariableSlot variableSlot = preferenceConstraint.getVariable();
+        ConstantSlot constantSlot = preferenceConstraint.getGoal();
+
+        BitVecExpr varBV = serializeVarSlot(variableSlot);
+        BitVecExpr constBV = serializeConstantSlot(constantSlot);
+
+        return context.mkEq(varBV, constBV);
+    }
+
+    class Z3SubtypeVariableCombos extends VariableCombos<SubtypeConstraint, BoolExpr> {
```

Comment:
I suggest separating the Z3___VariableCombos classes out into their own files in a package. This class is already pretty large.

Reply:
Sure. Will change this.

Reply:
Ah, I've recalled why I set them as inner classes. The reason of doing that is because `Z3**VariableCombos` is only being used by `Z3*Serializer` (this also is the purpose of VariableCombos in general). And also, those `Z3**VariableCombos` classes uses a lot fields and method in `Z3*Serializer`.

When I trying to separate those VarCombos out of the serializer class, I meet the problem of having to pass a reference of serializer to VarCombos, which seems a bit cumbersome.



I think there were too strong couplings between VarCombos and Serializers to separate them as top-level classes, just like the relationship between `AnnotatedTypeFactory` and `TreeAnnotator`.

Do you have any suggestions about how to deal this situation? 

Currently I cannot come up with a better idea than putting VarCombos as inner classes of Serializer.